Fix SIGSEGV in WebKit2 applications

vieiraa opened the pull request:
Detected this SIGSEGV when trying to login to game stores in Lutris.
May fix #175 
Also fixes crashes in Epiphany, and possibly in other WebKit2 applications as well.

mcatanzaro commented:
Can you please explain in your commit message what is going on and why your changes are correct? Particularly the commit "Only delete exported images." This code is confusing, and it's hard to reason about what's going on.

vieiraa commented:
Updated the code and the commit message. Figured that deleting images in the buffer callback was causing a double free, so I removed it. The code should be more clear now.

aperezdc reviewed (approved):
I finally got some time to try and understand this patch (sorry that it took a bit!) and it I think it makes sense. Let's merge it.

aperezdc commented:
Cherry-picked into the `wpebackend-fdo-1.12` branch as commits 8b1ea4db4d3821e80ec7ae0743ca8fb97f11662a and 6f9eebfd2d4e077cf80e9c56d1e874753468af0d.

alexgcastro commented:
The second patch in this pull requests is causing a leak in the gpu buffers, that ends up in random crashes. Could you provide some way to reproduce the double free described? That way we could provide a fix or we would have to revert until we check the original issue.

jameshilliard commented:
> The second patch in this pull requests is causing a leak in the gpu buffers, that ends up in random crashes. Could you provide some way to reproduce the double free described? That way we could provide a fix or we would have to revert until we check the original issue.

Well the issue showed up when running cog in valgrind like in #175.

alexgcastro commented:
> > The second patch in this pull requests is causing a leak in the gpu buffers, that ends up in random crashes. Could you provide some way to reproduce the double free described? That way we could provide a fix or we would have to revert until we check the original issue.
> 
> Well the issue showed up when running cog in valgrind like in #175.

I see, thanks. To me the problem in the valgrind looks like an invalid read of the image struct in some situations, but in some other situations is valid and we need that structure to delete the gpu buffer. Probably the fix needs to be to check why we call releaseImage when image struct is already freed, and if in that situation the gpu buffer is released. Also I would say we still have the invalid read in the `if (image->bufferResource)` because the image pointer is not correct.

The leak is in the gpu buffers which is out of the scope of valgrind, so probably after the patch even if valgrind does not report a leak it is there.

vieiraa commented:
IIRC, I could trigger a crash by just browsing on Epiphany (Gnome Web). I confirmed it was a double free on gdb.

vieiraa commented:
Perhaps a better fix would be marking the image for deletion on releaseImage (maybe using image->exported or creating a new flag) and actually deleting it on bufferDestroyListenerCallback.

psaavedra commented:
The `exported` safe-guards implements some relevant logic ifb897d2972a0bbc4d0157512f3da281ec7be917f


```
Use a destroy handle on the buffer wl_resource in order to properly track the 
resource's lifetime. In case the resource was destroyed while corresponding 
wl_resource was exported to the client, the subsequent release should avoid 
dispatching buffer release notification on the already-destroyed resource. 
```

psaavedra commented:
@vieiraa  dont you mind to test with this patch https://github.com/Igalia/WPEBackend-fdo/pull/178/files and verify if works for your case also?

vieiraa commented:
> @vieiraa dont you mind to test with this patch https://github.com/Igalia/WPEBackend-fdo/pull/178/files and verify if works for your case also?

Did some quick testing, and it does work.

alexgcastro commented:
We are back to this problem trying to review the new patch and we are trying to reproduce the problem, could someone give some more information about how to do it? We are trying with lutris.net, creating some account and logging in but all the combinations that we tried did not cause the crash and valgrind did not report the problem. Maybe it is because you were using some specific epiphany gtk4 version? Could you help reporting the exact command line, library versions and steps to reproduce it? It would be very helpful for us :pray: Thanks in advance :-)
